design/design-doc-observe-only-resources.md

```diff
@@ -0,0 +1,293 @@
+        metav1.TypeMeta   `json:",inline"`
+        metav1.ObjectMeta `json:"metadata,omitempty"`
+
++       // +kubebuilder:validation:XValidation:rule="self.managementPolicy == 'ObserveOnly' || has(self.forProvider.cidrBlock)",mess
+age="cidrBlock is a required parameter"
+        Spec   VPCSpec   `json:"spec"`
+        Status VPCStatus `json:"status,omitempty"`
+ }
+```
+
+## Future Work
+
+### Querying and Filtering
+
+Querying and filtering cloud resources is another common use case that could be relevant to making an observation. 
+Terraform uses [Data Sources] to observe existing resources by supporting some querying and filtering with a set of
+parameters specific to data source type. One can find and fetch data for the [most recent AMI] and a VPC with
+[desired tags].
+
+We will not support querying and filtering at *managed resources level* since it violates a fundamental principle with
+the managed resources, that is, having a one-to-one relationship between a managed resource and the external resource
+that it represents. When it comes to querying and filtering, it is possible that:
+
+- There are more than one matching resources
+- There are no matching resources
+- The matching resource may change in time, e.g., most-recent AMI
+
+Hence, we will leave implementing this functionality *in an upper layer*, which in turn will own managed resources with
+`managementPolicy: ObserveOnly`. In this model, it is totally fine if matching resources change in time, including
+having more than one or no matches where we would expect corresponding managed resources to come and go at runtime.
+
+We have two options to implement this functionality:
+
+**Option A: Introduce a new resource type `Query Resource`:**
+
+- This will no longer be a Managed Resource but a new type positioned on top of it, which will own and manage the
+lifecycle of Observe Only Managed Resources.
+- They will have their own kind and schema, e.g., to query/filter VPCs; we will have a `VPCQuery` resource.
+- Each provider implements Query Resources per type.
+- Leverages existing mechanisms in the provider (secret, IRSA, workload identity, etc.) to authenticate to the Cloud API.
+
+<img src="images/observe-only-query-resource.png" width="750" />
+
+
+**Option B: Defer this to the Composition layer, specifically, Compositions Functions:**
+
+- Compositions already operate as an upper layer by owning and managing the lifecycle of managed resources.
+- Querying and filtering are more like an imperative action that does not change the state of the external world and
+could be considered as part of auxiliary actions for compositing the infrastructure.
+- Authentication to the Cloud APIs is a problem that needs to be solved which is the biggest caveat of this approach.
+In the first pass of the composition functions design, even [passing sensitive configuration] to functions is not
+covered yet, and we would eventually need support for other authentication mechanisms.
```

Comment:
In some cases I could imagine stacking OO resources with Composition functions - e.g. have the Composition pass in a bunch of OO resources to a function which can then query them. I actually prefer this pattern because it keeps all communication with external systems flowing through cloud providers (keeping all auth etc in one place).

Reply:
Do you mean option A by 'prefer this pattern' ?

Reply:
No, something closer to option B.

However rather than a Composition Function talking directly to an external system (which I'd like to avoid as much as possible) I'm thinking that the Composition Function would "query" OO resources to derive information from them. I suppose a Composition Function pipeline could request the creation of OO resources that they need data from.

Consider for example an XR that needs to find an available subnet to use. I could imagine a pipeline of Composition Functions like:

1. Composition Function A requests an OO VPC resource be created.
2. Composition Function B reads the OO VPC resource to learn what subnets are in use, returns an available subnet.

Reply:
I see your point, and such a flow could help for certain use cases.
However, not sure how it could help for a use case like I need to get the latest AMI image.